diff --git a/include/trace/hooks/syscall_check.h b/include/trace/hooks/syscall_check.h
--- a/include/trace/hooks/syscall_check.h
+++ b/include/trace/hooks/syscall_check.h
@@ -1,7 +1,9 @@
 #ifndef _TRACE_HOOKS_SYSCALL_CHECK_H
 #define _TRACE_HOOKS_SYSCALL_CHECK_H
 
 #include <linux/tracepoint.h>
-#define TRACE_INCLUDE_PATH trace/hooks
+
+#ifndef TRACE_INCLUDE_PATH
+#define TRACE_INCLUDE_PATH trace/hooks
 #endif
 #define TRACE_INCLUDE_FILE syscall_check
 
 #undef TRACE_SYSTEM
